sematic/graph.py

```diff
@@ -0,0 +1,560 @@
+                        parent_future.nested_future = future
+
+                        if run.id in reset_run_ids:
+                            parent_future.state = FutureState.RAN
+
+                        return
+
+        def _clone_run(run_id: RunID):
+            run = self._runs_by_id[run_id]
+
+            kwargs, run_input_artifacts = _get_run_inputs(run_id)
+
+            func = run.get_func()
+
+            future = func(**kwargs)
+
+            input_artifacts[future.id] = run_input_artifacts
+
+            if run_id not in reset_run_ids:
+                value, output_artifact = _get_run_output(run_id)
+                if output_artifact is not None:
+                    output_artifacts[future.id] = output_artifact
+                    future.value = value
+
+            _set_parent_nested(future, run)
+
+            # Settings state for resolved runs unless reset
+            if FutureState[run.future_state] == FutureState.RESOLVED:  # type: ignore
+                if run.id not in reset_run_ids:
+                    future.state = FutureState.RESOLVED
+
+            futures_by_original_id[run.id] = future
+
+        for run_id in run_ids_by_execution_order:
+            if run_id in skip_run_ids:
+                continue
+
+            _clone_run(run_id)
+
+        if reset_from is None and len(futures_by_original_id) != len(self.runs):
+            raise RuntimeError("Not all futures duplicated")
+
+        # We return future sorted by how they would be sorted for a resolution
+        # from scratch: grouped by layer (outermost first), and sorter in reverse
+        # execution order within each layer
+        run_ids_by_reverse_execution_order = self._sorted_run_ids_by_layer(
+            run_sorter=self._reverse_execution_order
+        )
+
+        return (
+            collections.OrderedDict(
+                (
+                    (run_id, futures_by_original_id[run_id])
+                    for run_id in run_ids_by_reverse_execution_order
+                    if run_id in futures_by_original_id
+                )
+            ),
+            input_artifacts,
+            output_artifacts,
+        )
+
+
+EdgeFilterCallable = Callable[[List[Optional[RunID]], RunID], bool]
+
+
+def _upstream_edge_filter(
+    edges_by_destination_id: Dict[RunID, List[Edge]]
+) -> EdgeFilterCallable:
+    def _edge_filter(upstream_run_ids: List[Optional[RunID]], run_id: RunID):
+        return all(
+            edge.source_run_id in upstream_run_ids
+            for edge in edges_by_destination_id[run_id]
+        )
+
+    return _edge_filter
+
+
+def _downstream_edge_filter(
+    edges_by_source_id: Dict[RunID, List[Edge]]
+) -> EdgeFilterCallable:
+    def _edge_filter(downstream_run_ids: List[Optional[RunID]], run_id: RunID):
+        return all(
+            edge.destination_run_id in downstream_run_ids
+            for edge in edges_by_source_id[run_id]
+        )
+
+    return _edge_filter
+
+
+def _sort_layer_runs(
```

Comment:
It might be worth it (for testing, re-use, and readability, separation of concerns) to break out a `topological_sort` function with a signature like:

```
def topological_sort(ids: List[str], dependency_edges: List[Tuple[str, str]]) -> List[str]
```
Then you can define `_sort_layer_runs` as:
```python
def _sort_layer_runs(layer_run_ids: List[RunId], layer_edges: Edge, order: RunSortOrder):
    if order == RunSortOrder.FORWARD:
        edges_as_tuples = [(edge.source_run_id, edge.destination_run_id) for edge in layer_edges]
    else:
        edges_as_tuples = [(edge.destination_run_id, edge.source_run_id) for edge in layer_edges]
    edges_as_tuples = [edge for edge in edges_as_tuples if edge[0] is not None and edge[1] is not None]
    return topological_sort(layer_run_ids, edges_as_tuples)
```
The `topological_sort` alg can live in a separate util module.